Refactor vector editor properties

KoBeWi opened the pull request:
Follow-up to https://github.com/godotengine/godot/pull/75534#issuecomment-1549481806

The change I mentioned would require a tweak in all vector property classes *and I didn't feel like doing the same thing 4 times*, so I did a complete refactor. Instead of 6 classes that repeat literally the same thing there is only 1 now. The new class can handle any Vector with 2 or more components .-. Also, as a bonus, Vector4 now supports linking too.

Supersedes #75534
Closes #68165
Closes #73432

KoBeWi committed:
Refactor vector editor properties

KoBeWi commented on editor/editor_properties.cpp:
```diff
@@ -1689,93 +1689,155 @@ EditorPropertyEasing::EditorPropertyEasing() {
-		if (p_name == "x") {
-			spin[1]->set_value(spin[0]->get_value() * ratio_yx);
+		for (int i = 0; i < component_count; i++) {
+			if (p_name == COMPONENT_LABELS[i]) {
+				for (int j = 0; j < ratio.size(); j++) {
+					if (_get_ratio_component(j, 1) == i) {
+						for (int k = 0; k < component_count - 1; k++) {
+							spin_sliders[_get_ratio_component(j + k, 0)]->set_value_no_signal(spin_sliders[_get_ratio_component(j + k, 1)]->get_value() * ratio[j + k]);
+						}
+						break;
+					}
+				}
```
This needs to be simplified, but it took me too long to even figure it out ;_;
Maybe later >\_>

Geometror replied:
How about this? It's longer but easier to read (command-query separation and reduced the number of identation levels)
```suggestion

		// Find out which component has changed.
		int component_idx = -1;
		for (int i = 0; i < component_count; i++) {
			if (p_name == COMPONENT_LABELS[i]) {
				component_idx = i;
			}
		}

		int ratio_component_idx = -1;
		for (int j = 0; j < ratio.size(); j++) {
			int ratio_idx = _get_ratio_component(j, 1);
			if (ratio_idx == component_idx) {
				ratio_component_idx = j;
			}
		}

		// Update the other components based on the ratio.
		if (component_idx > 0 && ratio_component_idx > 0) {
			for (int k = 0; k < component_count - 1; k++) {
				int rc1 = _get_ratio_component(ratio_component_idx + k, 1);
				int rc2 = _get_ratio_component(ratio_component_idx + k, 0);
				double value = ratio[ratio_component_idx + k] * spin_sliders[rc1]->get_value();
				spin_sliders[rc2]->set_value_no_signal(value);
			}
		}
```

YuriSizov commented:
This needs a rebase.

KoBeWi commented:
Rebased. I'm not convinced about the suggestion. It makes the code easier to read, but doesn't simplify the logic, which is convoluted and inefficient. I'll try to do something about it, but it's also something that can be done in a follow-up PR. For now I left a TODO comment.

The exact problem is that ratios are packed in an array that looks something like `[ xy, yx, zy, yz, zx, zy]` (I don't remember the exact order now xd), so it needs some advanced index magic to correctly map it. Of course the order is arbitrary, I just crammed all ratios into a single array and needed some way to do so. Maybe it can be changed to a HashMap or something.

kleonc commented:
> The exact problem is that ratios are packed in an array that looks something like `[ xy, yx, zy, yz, zx, zy]` (I don't remember the exact order now xd), so it needs some advanced index magic to correctly map it. Of course the order is arbitrary

@KoBeWi You can try something like this:
```
int n = component_count
int m = n - 1

// Updating sliders.
// slider[i].value = slider[j].value * ratios[k]
// (calculating k from i, j)
int j = changed_value_index
for i in n:
	if i == j:
		continue
	int k = m * j + ((i - j + n) % n) - 1
	slider[i].value = slider[j].value * ratio[k]

// Updating ratios.
// ratios[k] = slider[i].value / slider[j].value
// (calculating i, j from k)
for k in ratios.size(): // ratios.size() == n * m
	int j = k / m
	int i = (j + 1 + (k % m)) % n
	ratios[k] = slider[i].value / slider[j].value
```
It should result in ratios orderings (`yx` means `y / x`):
- `[yx,  xy]`,
- `[yx, zx,  zy, xy,  xz, yz]`,
- `[yx, zx, wx,  zy, wy, xy,  wz, xz, yz,  xw, yw, zw]`.

YuriSizov commented on editor/editor_properties.cpp:
```diff
@@ -1806,31 +1874,42 @@ EditorPropertyVector2::EditorPropertyVector2(bool p_force_wide) {
+	ratio.fill(1.0);
+
 	linked = memnew(TextureButton);
 	linked->set_toggle_mode(true);
 	linked->set_stretch_mode(TextureButton::STRETCH_KEEP_CENTERED);
-	linked->connect(SNAME("pressed"), callable_mp(this, &EditorPropertyVector2::_update_ratio));
+	linked->connect(SNAME("pressed"), callable_mp(this, &EditorPropertyVectorN::_update_ratio));
+	linked->connect(SNAME("toggled"), callable_mp(this, &EditorPropertyVectorN::_store_link));
 	hb->add_child(linked);
```
I think that the button is missing a tooltip, preferably a dynamic one corresponding to its current state or reflecting what's going to happen when you click on it.

YuriSizov commented:
Something is wrong with linked values. I'm not sure if it's new to this PR or if it's a pre-existing issue, but it's an issue nonetheless. If you have a vector with components which have different signs, things get weird.

Sometimes the value and the sign of each component get in sync:


https://github.com/godotengine/godot/assets/11782833/b5c9b04e-0895-465d-8976-0ba2872149c6


https://github.com/godotengine/godot/assets/11782833/35b9d12a-2994-4d53-b297-f1721036d6e0


And sometimes it's only the values, but the signs are appropriately different:


https://github.com/godotengine/godot/assets/11782833/09de4cce-5012-4f4c-9aeb-bf48f2bf019f


It's a bit harder to test with non-integer vectors, as they are much slower to change, but they are not immune to the issue:


https://github.com/godotengine/godot/assets/11782833/ee5940b9-c2bd-4b3b-a7d7-b86984bc4250


-----

Note that it doesn't matter which component you drag, though initially I thought it did matter. What matters is how many times you cross the 0 threshold, and how the big numbers are (the bigger, the longer it takes them to sync).

YuriSizov commented:
Other than that bug, everything seems to be working fine, and I'm sure that code is way better to maintain now :)

YuriSizov commented:
Okay, a correction. I only noticed this with negative numbers at first, so I assumed it was related, but indeed the only problem is that I was crossing the negative-positive boundary. The issue is repeatable even if you start with both positive numbers but a ratio different from 1:1. Crossing the 0 value with one of the components (maybe a few times) breaks the ratio.

KoBeWi says this is because when 0 is involved we can't have a valid ratio between components, which is reasonable. Perhaps we could cache the last valid ratio and use that?

kleonc commented:
@YuriSizov Currently in `EditorPropertyVectorN::_update_ratio` if any of the sliders has zero value then all ratios will be set to `1.0`. I think this should be changed and be resolved per component.

When dragging/changing value of a non-zero component, then all components can be scaled properly according to the ratios (at the start of the change), including the zero ones (they'll remain zero). So for `(x, y, z)` changing e.g. non-zero `x` to `new_x` would result in `new_x * (1, y/x, z/x)`. Like for `(1, 0, 2)` changing `x` to `3` would result in `(3, 0, 6)`. That's correct and expected and should be made like that.

The question is what the behavior should be when dragging/changing value of a zero component. I don't think there's an obvious expected behavior, it's rather a matter of making an arbitrary choice.
- According to the ratios what would be correct is setting other non-zero components to infinities. It might be not desired though. :thinking: E.g. for `(1, 0, -2, 0)` changing `y` to `3` would result in `(INF, 3, -INF, 3)` (could be argued it should result in`(INF, 3, -INF, 0)` instead).
- Another possibility would be to totally ignore linking for zero values. So for `(1, 0, -2, 0)` changing `y` to `3` would result in `(1, 3, -2, 0)` (or keep only the zeros linked so result would be `(1, 3, -2, 3)`:laughing:).
- Maybe some other possibilities...

The decision on the behavior is needed before the implementation of course. :upside_down_face:

Also it should be ensured that e.g. when dragging the value in the inspector the `_update_ratio` is called only when the drag starts/ends. Haven't checked if it's currently like that but calling it in the middle of the drag could likely cause issues when going through zero (because of the different behavior), like the one mentioned in https://github.com/godotengine/godot/pull/77135#issuecomment-1568379310.

YuriSizov commented:
I like what you suggest and I like the idea of ignoring the ratio when starting the change from a zeroed component.

KoBeWi commented:
I added tooltip and moved the properties to a new file.
I still did not touch the ratio logic. The bug you mentioned already existed before and this PR makes it easier to fix (as you need to do it only once xd).

YuriSizov commented:
I opened a report for that problem: https://github.com/godotengine/godot/issues/77687

YuriSizov commented:
Thanks!
